command/agent/srv.go

```diff
@@ -0,0 +1,114 @@
+package agent
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"time"
+)
+
+const (
+	srvPollInterval = 60 * time.Second
+	srvFindTimeout  = srvPollInterval / 2
+)
+
+// AgentSRV periodically polls an SRV record
+// And attempts to join the hosts supplied
+type AgentSRV struct {
+	agent      *Agent
+	srvrecords []string
+	logger     *log.Logger
+	replay     bool
+}
+
+// NewAgentSRV is used to create a new AgentSRV
+func NewAgentSRV(agent *Agent, logOutput io.Writer, replay bool, srvrecords []string) (*AgentSRV, error) {
+
+	// Initialize the AgentSRV
+	m := &AgentSRV{
+		agent:      agent,
+		srvrecords: srvrecords,
+		logger:     log.New(logOutput, "", log.LstdFlags),
+		replay:     replay,
+	}
+
+	// Start the poller in the background
+	go m.run()
+	return m, nil
+}
+
+// run is a long running goroutine that scans for new hosts periodically
+func (m *AgentSRV) run() {
+
+	for {
+		// Set up a channel and select so we can timeout if findSRV takes too long
+		// set channel to 1 so that goroutines can't pile-up
+		c := make(chan []string, 1)
+		go func() { c <- m.findSRV() }()
+		select {
+		case records := <-c:
+			// Attempt the join only if there are new records
+			if len(records) > 0 {
+				n, err := m.agent.Join(records, m.replay)
+
+				if err != nil {
+					m.logger.Printf("[ERR] agent.srv: Failed to join: %v", err)
+				}
+				if n > 0 {
+					m.logger.Printf("[INFO] agent.srv: Joined %d hosts", n)
+				}
+			}
+		// Report the timeout
+		case <-time.After(srvFindTimeout):
+			m.logger.Printf("[ERR] agent.srv: findSRV timed out")
+		}
+		// Sleep until it's time to poll again
+		time.Sleep(srvPollInterval)
+	}
+}
+
+// findSRV looks up the SRV records and returns a slice of all SRV records
+// that are not currently cluster members
+func (m *AgentSRV) findSRV() []string {
+	var hosts []string
+
+	// map the members so that we only do a single O(n) search through members
+	known_members := make(map[string]bool)
```

Comment:
members is not already a map, it's an array - which is the problem here. In order to save searching through the list over and over, I do a one time mapping (O(n)) to a map so that I can get O(1) lookups.

https://github.com/hashicorp/serf/blob/master/serf/serf.go#L696-L697

Unless I'm misunderstanding you, and there is a way to get this as a map?

```
// Members returns a point-in-time snapshot of the members of this cluster.
func (s *Serf) Members() []Member {
```